### PR TITLE
Move most stuff to a lib

### PR DESCRIPTION
**Pull request by designatednerd**
#### Enhancements
- Created a `SourceDocsLib` which can be imported separately from the executable, and used within Swift Package Manager Scripts
- Made a bunch of stuff public to facilitate this

#### Bug Fixes
- Reset  the markdown index when generation exits to allow multiple generation runs in one go without all targets getting all markdown
- Enabled  the access level tests

#### Further comments 
- For some weird reason this is only working with my project when I do `swift run MyProject` and not when I build my project in Xcode. When I build in Xcode i'm getting `Please confirm that `xcodebuild` is building a Swift module.`. I *think* from looking at a bunch of issues on Jazzy, this is probably related to SourceKitten, which they also depend on. However, I think this is at least workable in terms of running on the command line similarly to the previous version. 
- Tests are passing when run via `swift test` but will fail if run via Xcode - the `system` lib seems to not really enjoy being run from Xcode. 
- `Commandant` and a bunch of other stuff being used here just pipe everything out to `stdout`, which makes it awfully hard to test. I know there's [`ShellOut`](https://github.com/JohnSundell/ShellOut) that returns strings in addition to sending them to the console, which makes it a LOT easier to test that things are working correctly. 

I'm good with this getting merged as a phase one since it works, but wanted to share those notes for use in terms of future rethinks.

### CODE REVIEW

**Comment by designatednerd** [comment]
Addresses #39

**Comment by designatednerd** [comment]
Poked Travis settings to use Swift 5.1, the only thing still failing is `swiftlint` because it's on `strict` - didn't want to turn that off without checking first, but that does make it fail the build even for warnings. Should I make alterations or turn that off?

**Comment by eneko** [comment]
Hi @designatednerd, I'm looking into this (almost two months later, I know 🙈).

First of all, thank you very much for doing this. I think it's a great change for the repo.

Few changes I'd like to do before merging:
- Looks like a lot of the command handling code has been moved to the library. While the core logic should be moved, command-related code should be for the command line tool only.

I'm going to try to push this changes, but never done that on a fork PR before, so not sure how it works. Will see.

**Comment by eneko** [comment]
Looks like I cannot push to this PR, so a new one has been created: https://github.com/eneko/SourceDocs/pull/41

**Comment by designatednerd** [comment]
So the problem with moving the command-handling code back to the library is that I can't actually call any of the commands outside the library if we do that.  [Here's how I'm using it now](https://github.com/apollographql/apollo-ios/blob/master/SwiftScripts/Sources/DocumentationGenerator/main.swift#L44). If there's a way to separate the options stuff from the commander stuff, that'd probably be best, but I didn't see a straightforward way to do that. 

Basically this isn't all that helpful if I can't run the commands from outside the CLI 🙃

**Comment by eneko** [comment]
I see, thanks for the link. Yeah, a new public interface will have to be defined. The idea is that the library part would not need to depend on Commandant, Rainbow, or any other command line related code.

I can take care of that, not sure how soon but will work towards it. Will ping you when I have something ready! :)

**Comment by eneko** [comment]
Hi @designatednerd,

I've submitted a new pull request to move code into `SourceDocsLib`. Please check out #42 and let me know your thoughts.

You would need to update your code in apollo-ios like this:

```swift
let options = DocumentOptions(moduleName: target.name,
                              linkEndingText: "/",
                              inputFolder: sourceRootURL.path,
                              outputFolder: outputURL.path,
                              clean: true,
                              xcodeArguments: [
                                "-scheme",
                                target.scheme,
                                "-project",
                                "Apollo.xcodeproj"])

do {
  try DocumentationGenerator(options: options).run()
  CodegenLogger.log("Generated docs for \(target.name)")
} catch {
  CodegenLogger.log("Error generating docs for \(target.name): \(error)", logLevel: .error)
  exit(1)
}
```

**Comment by eneko** [comment]
Closed in favor of #42
